### ISSUE TITLE
CLI sends large numbers in scientific notation in param values

### ISSUE THREAD

**Issue by TheSimpleZ** [argoproj/argo#3990]
## Summary 

When running the [arguments parameter example](https://github.com/argoproj/argo/blob/master/examples/arguments-parameters.yaml) using the CLI, any large numbers gets converted to scientific notation before sent in. 

20200219 becomes 2.0200219e+07. If I submit the workflow through the UI everything works fine. 

I would expect it to deliver the parameter as-is. 

It doesn't matter whether i use the `-p` flag to set the parameter value or if i set it directly in the workflow yaml. If I use the `argo submit` command, the value gets converted to scientific notation. 

## Diagnostics

What version of Argo Workflows are you running?

v2.10.1

Workflow: https://github.com/argoproj/argo/blob/master/examples/arguments-parameters.yaml

Running with : `argo submit -p message=20200219 https://raw.githubusercontent.com/argoproj/argo/master/examples/arguments-parameters.yaml`

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  creationTimestamp: "2020-09-10T14:12:31Z"
  generateName: arguments-parameters-
  generation: 4
  labels:
    workflows.argoproj.io/completed: "true"
    workflows.argoproj.io/phase: Error
  name: arguments-parameters-xkvrc
  namespace: default
  resourceVersion: "24479158"
  selfLink: /apis/argoproj.io/v1alpha1/namespaces/default/workflows/arguments-parameters-xkvrc
  uid: 7deb13d8-2c2f-4d47-9430-22774d2ee3db
spec:
  arguments:
    parameters:
    - name: message
      value: "2.0200219e+07"
  entrypoint: whalesay
  templates:
  - arguments: {}
    container:
      args:
      - '{{inputs.parameters.message}}'
      command:
      - cowsay
      image: docker/whalesay:latest
      name: ""
      resources: {}
    inputs:
      parameters:
      - name: message
    metadata: {}
    name: whalesay
    outputs: {}
status:
  conditions:
  - status: "True"
    type: Completed
  finishedAt: "2020-09-10T14:12:35Z"
  message: 'failed to save outputs: Failed to establish pod watch: unknown (get pods)'
  nodes:
    arguments-parameters-xkvrc:
      displayName: arguments-parameters-xkvrc
      finishedAt: "2020-09-10T14:12:34Z"
      id: arguments-parameters-xkvrc
      inputs:
        parameters:
        - name: message
          value: "2.0200219e+07"
      message: 'failed to save outputs: Failed to establish pod watch: unknown (get
        pods)'
      name: arguments-parameters-xkvrc
      phase: Error
      startedAt: "2020-09-10T14:12:31Z"
      templateName: whalesay
      type: Pod
  phase: Error
  startedAt: "2020-09-10T14:12:31Z"
```

```
time="2020-09-10T14:12:31Z" level=info msg="Processing workflow" namespace=default workflow=arguments-parameters-xkvrc
time="2020-09-10T14:12:31Z" level=info msg="Updated phase  -> Running" namespace=default workflow=arguments-parameters-xkvrc
time="2020-09-10T14:12:31Z" level=info msg="Pod node {arguments-parameters-xkvrc arguments-parameters-xkvrc arguments-parameters-xkvrc Pod whalesay nil    Pending   2020-09-10 14:12:31.484613759 +0000 UTC 0001-01-01 00:00:00 +0000 UTC   <nil> nil nil [] []} initialized Pending" namespace=default workflow=arguments-parameters-xkvrc
time="2020-09-10T14:12:31Z" level=info msg="Created pod: arguments-parameters-xkvrc (arguments-parameters-xkvrc)" namespace=default workflow=arguments-parameters-xkvrc
time="2020-09-10T14:12:31Z" level=info msg="Workflow update successful" namespace=default phase=Running resourceVersion=24479127 workflow=arguments-parameters-xkvrc
time="2020-09-10T14:12:32Z" level=info msg="Processing workflow" namespace=default workflow=arguments-parameters-xkvrc
time="2020-09-10T14:12:32Z" level=info msg="Updating node &NodeStatus{ID:arguments-parameters-xkvrc,Name:arguments-parameters-xkvrc,DisplayName:arguments-parameters-xkvrc,Type:Pod,TemplateName:whalesay,TemplateRef:nil,Phase:Pending,BoundaryID:,Message:,StartedAt:2020-09-10 14:12:31 +0000 UTC,FinishedAt:0001-01-01 00:00:00 +0000 UTC,PodIP:,Daemoned:nil,Inputs:&Inputs{Parameters:[]Parameter{Parameter{Name:message,Default:nil,Value:*2.0200219e+07,ValueFrom:nil,GlobalName:,},},Artifacts:[]Artifact{},},Outputs:nil,Children:[],OutboundNodes:[],StoredTemplateID:,WorkflowTemplateName:,TemplateScope:,ResourcesDuration:ResourcesDuration{},} message: ContainerCreating"
time="2020-09-10T14:12:32Z" level=info msg="Workflow update successful" namespace=default phase=Running resourceVersion=24479140 workflow=arguments-parameters-xkvrc
time="2020-09-10T14:12:33Z" level=info msg="Processing workflow" namespace=default workflow=arguments-parameters-xkvrc
time="2020-09-10T14:12:35Z" level=info msg="Processing workflow" namespace=default workflow=arguments-parameters-xkvrc
time="2020-09-10T14:12:35Z" level=info msg="Updating node &NodeStatus{ID:arguments-parameters-xkvrc,Name:arguments-parameters-xkvrc,DisplayName:arguments-parameters-xkvrc,Type:Pod,TemplateName:whalesay,TemplateRef:nil,Phase:Pending,BoundaryID:,Message:ContainerCreating,StartedAt:2020-09-10 14:12:31 +0000 UTC,FinishedAt:0001-01-01 00:00:00 +0000 UTC,PodIP:,Daemoned:nil,Inputs:&Inputs{Parameters:[]Parameter{Parameter{Name:message,Default:nil,Value:*2.0200219e+07,ValueFrom:nil,GlobalName:,},},Artifacts:[]Artifact{},},Outputs:nil,Children:[],OutboundNodes:[],StoredTemplateID:,WorkflowTemplateName:,TemplateScope:,ResourcesDuration:ResourcesDuration{},} status Pending -> Error"
time="2020-09-10T14:12:35Z" level=info msg="Updating node &NodeStatus{ID:arguments-parameters-xkvrc,Name:arguments-parameters-xkvrc,DisplayName:arguments-parameters-xkvrc,Type:Pod,TemplateName:whalesay,TemplateRef:nil,Phase:Error,BoundaryID:,Message:,StartedAt:2020-09-10 14:12:31 +0000 UTC,FinishedAt:0001-01-01 00:00:00 +0000 UTC,PodIP:,Daemoned:nil,Inputs:&Inputs{Parameters:[]Parameter{Parameter{Name:message,Default:nil,Value:*2.0200219e+07,ValueFrom:nil,GlobalName:,},},Artifacts:[]Artifact{},},Outputs:nil,Children:[],OutboundNodes:[],StoredTemplateID:,WorkflowTemplateName:,TemplateScope:,ResourcesDuration:ResourcesDuration{},} message: failed to save outputs: Failed to establish pod watch: unknown (get pods)"
time="2020-09-10T14:12:35Z" level=info msg="Updated phase Running -> Error" namespace=default workflow=arguments-parameters-xkvrc
time="2020-09-10T14:12:35Z" level=info msg="Updated message  -> failed to save outputs: Failed to establish pod watch: unknown (get pods)" namespace=default workflow=arguments-parameters-xkvrc
time="2020-09-10T14:12:35Z" level=info msg="Marking workflow completed" namespace=default workflow=arguments-parameters-xkvrc
time="2020-09-10T14:12:35Z" level=info msg="Checking daemoned children of " namespace=default workflow=arguments-parameters-xkvrc
time="2020-09-10T14:12:35Z" level=info msg="Workflow update successful" namespace=default phase=Error resourceVersion=24479158 workflow=arguments-parameters-xkvrc
time="2020-09-10T14:12:36Z" level=info msg="Labeled pod default/arguments-parameters-xkvrc completed"

```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by alexec** [comment]
@TheSimpleZ do you know if this is a new issue? E.g. appeared in v2.10?

**Comment by TheSimpleZ** [comment]
No idea. I just recently started using Argo.

**Comment by sarabala1979** [comment]
it works for me. Can you provide your CLI version? `argo version`

```
argo: v2.9.4+20d2ace.dirty
  BuildDate: 2020-07-25T08:30:49Z
  GitCommit: 20d2ace3d5344db68ce1bc2a250bbb1ba9862613
  GitTreeState: dirty
  GitTag: v2.9.4
  GoVersion: go1.14.5
  Compiler: gc
  Platform: darwin/amd64
```

```
 argo submit examples/hello-world.yaml  -p message=202002192
```

```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  creationTimestamp: "2020-09-11T06:05:58Z"
  generateName: hello-world-
  generation: 4
  labels:
    workflows.argoproj.io/archive-strategy: "false"
    workflows.argoproj.io/completed: "true"
    workflows.argoproj.io/phase: Succeeded
  managedFields:
  - apiVersion: argoproj.io/v1alpha1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:generateName: {}
        f:labels:
          .: {}
          f:workflows.argoproj.io/archive-strategy: {}
      f:spec:
        .: {}
        f:arguments:
          .: {}
          f:parameters: {}
        f:entrypoint: {}
        f:templates: {}
      f:status: {}
    manager: argo
    operation: Update
    time: "2020-09-11T06:05:58Z"
  - apiVersion: argoproj.io/v1alpha1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:labels:
          f:workflows.argoproj.io/completed: {}
          f:workflows.argoproj.io/phase: {}
      f:status:
        f:conditions: {}
        f:finishedAt: {}
        f:nodes:
          .: {}
          f:hello-world-dghsq:
            .: {}
            f:displayName: {}
            f:finishedAt: {}
            f:hostNodeName: {}
            f:id: {}
            f:name: {}
            f:outputs:
              .: {}
              f:artifacts: {}
              f:exitCode: {}
            f:phase: {}
            f:resourcesDuration:
              .: {}
              f:cpu: {}
              f:memory: {}
            f:startedAt: {}
            f:templateName: {}
            f:templateScope: {}
            f:type: {}
        f:phase: {}
        f:resourcesDuration:
          .: {}
          f:cpu: {}
          f:memory: {}
        f:startedAt: {}
    manager: ___controller
    operation: Update
    time: "2020-09-11T06:06:05Z"
  name: hello-world-dghsq
  namespace: argo
  resourceVersion: "54262"
  selfLink: /apis/argoproj.io/v1alpha1/namespaces/argo/workflows/hello-world-dghsq
  uid: b3ba1d85-5c81-4719-bc8a-b07a74b287a6
spec:
  arguments:
    parameters:
    - name: message
      value: "202002192"
  entrypoint: whalesay
  templates:
  - arguments: {}
    container:
      args:
      - hello world
      command:
      - cowsay
      image: docker/whalesay:latest
      name: ""
      resources: {}
    inputs: {}
    metadata: {}
    name: whalesay
    outputs: {}
status:
  conditions:
  - status: "True"
    type: Completed
  finishedAt: "2020-09-11T06:06:05Z"
  nodes:
    hello-world-dghsq:
      displayName: hello-world-dghsq
      finishedAt: "2020-09-11T06:06:04Z"
      hostNodeName: kind-control-plane
      id: hello-world-dghsq
      name: hello-world-dghsq
      outputs:
        artifacts:
        - archiveLogs: true
          name: main-logs
          s3:
            accessKeySecret:
              key: accesskey
              name: my-minio-cred
            bucket: my-bucket
            endpoint: minio:9000
            insecure: true
            key: hello-world-dghsq/hello-world-dghsq/main.log
            secretKeySecret:
              key: secretkey
              name: my-minio-cred
        exitCode: "0"
      phase: Succeeded
      resourcesDuration:
        cpu: 5
        memory: 3
      startedAt: "2020-09-11T06:05:58Z"
      templateName: whalesay
      templateScope: local/hello-world-dghsq
      type: Pod
  phase: Succeeded
  resourcesDuration:
    cpu: 5
    memory: 3
  startedAt: "2020-09-11T06:05:58Z"

```

**Comment by TheSimpleZ** [comment]
I think I'm using the latest version :)

```
> argo version
argo: v2.10.1
  BuildDate: 2020-09-02T22:55:38Z
  GitCommit: 854444e47ac00d146cb83d174049bfbb2066bfb2
  GitTreeState: clean
  GitTag: v2.10.1
  GoVersion: go1.13.4
  Compiler: gc
  Platform: windows/amd64
``´

**Comment by TheSimpleZ** [comment]
As i stated in the original post, if i modify the workflow and i add the value in the YAML directly it behaves the same. 

```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: arguments-parameters-
spec:
  entrypoint: whalesay
  # Parameters can be passed/overridden via the argo CLI.
  # To override the printed message, run `argo submit` with the -p option:
  # $ argo submit examples/arguments-parameters.yaml -p message="goodbye world"
  arguments:
    parameters:
    - name: message
      value: 20200219 # Becomes 2.0200219e+07

  templates:
  - name: whalesay
    inputs:
      parameters:
      - name: message
    container:
      image: docker/whalesay:latest
      command: [cowsay]
      args: ["{{inputs.parameters.message}}"]
```
However, if I add quotes around the number, it's passed correctly. It seems like if the value if the value is considered an integer, it will change it to scientific notation. So forcing it to be a string seems to help.

```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: arguments-parameters-
spec:
  entrypoint: whalesay
  # Parameters can be passed/overridden via the argo CLI.
  # To override the printed message, run `argo submit` with the -p option:
  # $ argo submit examples/arguments-parameters.yaml -p message="goodbye world"
  arguments:
    parameters:
    - name: message
      value: "20200219" # Becomes 20200219

  templates:
  - name: whalesay
    inputs:
      parameters:
      - name: message
    container:
      image: docker/whalesay:latest
      command: [cowsay]
      args: ["{{inputs.parameters.message}}"]
```

However, I don't know of any way of doing the same thing from the command line. Simply running `argo submit -p message="20200219" https://raw.githubusercontent.com/argoproj/argo/master/examples` did not work 😕

**Comment by alexec** [comment]
I'm betting this is fixed by #3963

**Comment by KevinGG** [comment]
I'm using argo: `v2.11.0-rc2`. I also see the issue and try to work around it with `-f` option.

I'm using the hello world example.
```
curl -sLO https://raw.githubusercontent.com/argoproj/argo/master/examples/hello-world.yaml
```
And I tried passing parameters from a file `test.yaml`
```
message: "1599844140"
```
Then
```
argo submit -n argo -f test.yaml --watch hello-world.yaml
```
The parameter becomes
```
message:           {1 0 1.59984414e+09}
```
The string defined in the input `test.yaml` file is treated as an integer and converted to a scientific notation.
I assume this https://github.com/argoproj/argo/issues/1894 caused the bug. Because it would pass literal quotes before that fix, though neither behavior is desired.

I understand it might be hard to determine the parameter type from the cli in bash, but the input yaml file approach should be able to give enough information to the cli about the types of parameters.
